cmd/gateway-sia-cachelayer.go

```diff
@@ -0,0 +1,1194 @@
+		UploadCheckFreqMs: 3000,
+		MaxCacheSizeBytes: 10000000000,
+		CacheTicker:       nil,
+		Db:                nil,
+		DbMutex:           &sync.Mutex{},
+	}, nil
+}
+
+// Start will start running the Cache Layer
+func (cache *SiaCacheLayer) Start() *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.Start")
+	cache.loadSiaEnv()
+
+	cache.DbMutex = &sync.Mutex{}
+
+	cache.ensureCacheDirExists()
+
+	// Open and initialize database
+	err := cache.initDatabase()
+	if err != nil {
+		return err
+	}
+
+	// Start the cache management process
+	cache.CacheTicker = time.NewTicker(time.Second * time.Duration(cache.ManagerDelaySec))
+	go func() {
+		for _ = range cache.CacheTicker.C {
+			cache.manager()
+		}
+	}()
+
+	return nil
+}
+
+// Attempt to load Sia config from ENV
+func (cache *SiaCacheLayer) loadSiaEnv() {
+	tmp := os.Getenv("SIA_MANAGER_DELAY_SEC")
+	if tmp != "" {
+		i, err := strconv.ParseInt(tmp, 10, 64)
+		if err == nil {
+			cache.ManagerDelaySec = i
+		}
+	}
+
+	tmp = os.Getenv("SIA_UPLOAD_CHECK_FREQ_MS")
+	if tmp != "" {
+		i, err := strconv.ParseInt(tmp, 10, 64)
+		if err == nil {
+			cache.UploadCheckFreqMs = i
+		}
+	}
+
+	tmp = os.Getenv("SIA_CACHE_MAX_SIZE_BYTES")
+	if tmp != "" {
+		i, err := strconv.ParseInt(tmp, 10, 64)
+		if err == nil {
+			cache.MaxCacheSizeBytes = i
+		}
+	}
+
+	if cache.DebugMode {
+		fmt.Printf("SIA_CACHE_MAX_SIZE_BYTES: %d\n", cache.MaxCacheSizeBytes)
+		fmt.Printf("SIA_MANAGER_DELAY_SEC: %d\n", cache.ManagerDelaySec)
+		fmt.Printf("SIA_UPLOAD_CHECK_FREQ_MS: %d\n", cache.UploadCheckFreqMs)
+	}
+}
+
+// Stop will stop the SiaCacheLayer
+func (cache *SiaCacheLayer) Stop() {
+	cache.debugmsg("SiaCacheLayer.Stop")
+	// Stop cache management process
+	cache.CacheTicker.Stop()
+
+	// Close the database
+	cache.lockDB()
+	cache.Db.Close()
+	cache.unlockDB()
+}
+
+func (cache *SiaCacheLayer) lockDB() {
+	//cache.debugmsg("SiaCacheLayer.LockDB")
+	cache.DbMutex.Lock()
+}
+
+func (cache *SiaCacheLayer) unlockDB() {
+	//cache.debugmsg("SiaCacheLayer.UnlockDB")
+	cache.DbMutex.Unlock()
+}
+
+// InsertBucket will attempt to insert a new bucket
+func (cache *SiaCacheLayer) InsertBucket(bucket string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.InsertBucket")
+
+	cache.lockDB()
+	defer cache.unlockDB()
+
+	stmt, err := cache.Db.Prepare("INSERT INTO buckets(name, created, policy) values(?,?,?)")
+	if err != nil {
+		return siaErrorDatabaseInsertError
+	}
+
+	defer stmt.Close()
+
+	policyInfo := policy.BucketAccessPolicy{}
+	res, _ := json.Marshal(&policyInfo)
+	_, err = stmt.Exec(bucket, time.Now().Unix(), string(res))
+	if err != nil {
+		cache.debugmsg(err.Error())
+		return siaErrorDatabaseInsertError
+	}
+
+	return nil
+}
+
+// DeleteBucket will attempt to delete an existing bucket
+func (cache *SiaCacheLayer) DeleteBucket(bucket string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.DeleteBucket")
+	// First, should delete all objects in the bucket
+	objects, err := cache.ListObjects(bucket)
```

Comment:
Delete bucket shouldn't delete all top level objects.. only multipart i.e temporary objects can be deleted. Otherwise DeleteBucket() should return ENOTEMPTY.  Look at the fs-v1.go implementation for more inputs.